subprojects/core/src/main/java/org/gradle/api/internal/instantiation/ParamsMatchingConstructorSelector.java

```diff
@@ -49,6 +50,10 @@ public SelectedConstructor forParams(final Class<?> type, Object[] params) {
             }
         });
 
+        if (constructors.size() == 1) {
+            return constructors.get(0);
+        }
```

Comment:
I see the optimization here, but aren't you bypassing all the validation of ensuring that the `params` match here? Couldn't you end up with other weird errors later because of this?